tests/unit/machine/all_machines_unittest.cc

```diff
@@ -0,0 +1,386 @@
+
+		auto data = generate_data(machine);
+
+		get_global_parallel()->set_num_threads(1);
+		machine->set_labels(data.second);
+		if (machine->has("seed"))
+			machine->put("seed", 1);
+		machine->train(data.first);
+		auto result_single = machine->apply(data.first);
+
+		init_machine(machine);
+		get_global_parallel()->set_num_threads(4);
+		machine2->set_labels(data.second);
+		if (machine2->has("seed"))
+			machine2->put("seed", 1);
+		machine2->train(data.first);
+		auto result_multi = machine2->apply(data.first);
+
+		EXPECT_TRUE(result_single->equals(result_multi));
+	}
+}
+
+
+TEST(AllMachines, view_subsampling_consistency)
+{
+	std::set<std::string> ignores = {
+			"RandomForest" // BaggingMachine.cpp line 153: Number of bag is not set!"
+			};
+	for (auto obj : sg_object_iterator<untemplated_sgobject>(all_machines).ignore(ignores))
+	{
+		auto machine = obj->as<CMachine>();
+		SCOPED_TRACE(machine->get_name());
+
+		init_machine(machine);
+		auto data = generate_data(machine);
+
+		auto X = data.first->get_feature_matrix();
+		auto y = data.second->get_labels();
+
+		SGVector<index_t> subset = {1,3,4,6};
+
+		auto features_subset = view(data.first, subset);
+		auto labels_subset = view(data.second, subset);
+
+		SGMatrix<float64_t> X_subsampled(X.num_rows, subset.size());
+		SGVector<float64_t> y_subsampled(subset.size());
+
+		for (auto i : range(subset.size()))
+		{
+			memcpy(X_subsampled.get_column_vector(i), X.get_column_vector(subset[i]), X.num_rows * sizeof(decltype(X(0,0))));
+			y_subsampled[i] = y[subset[i]];
+		}
+
+		auto features_subsampled = new CDenseFeatures<float64_t>(X_subsampled);
+		auto labels_subsampled = make_clone(data.second);
+		labels_subsampled->set_labels(y_subsampled);
+
+		auto machine_subset = make_clone(machine);
+		auto machine_subsampled = make_clone(machine);
+		if (machine->has("seed"))
+		{
+			machine_subset->put("seed", 1);
+			machine_subsampled->put("seed", 1);
+		}
+
+		machine_subset->set_labels(labels_subset);
+		machine_subset->train(features_subset);
+		machine_subsampled->set_labels(labels_subsampled);
+		machine_subsampled->train(features_subsampled);
+
+		auto result_subset = machine_subset->apply(data.first);
+		auto result_subsampled = machine_subsampled->apply(data.first);
+
+		EXPECT_TRUE(result_subset->equals(result_subsampled));
+	}
+}
+
+TEST(AllMachines, cv_thread_consistency)
+{
+	std::set<std::string> ignores = {
+			"RandomForest" // BaggingMachine.cpp line 153: Number of bag is not set!"
+			};
+	for (auto obj : sg_object_iterator<untemplated_sgobject>(all_machines).ignore(ignores))
+	{
+		auto machine = obj->as<CMachine>();
+		SCOPED_TRACE(machine->get_name());
+
+		init_machine(machine);
+		auto data = generate_data(machine);
+		auto cv = generate_cv(machine, data);
+		auto cv2 = make_clone(cv);
+
+		get_global_parallel()->set_num_threads(1);
+		cv->put("seed", 1);
+		auto result_single = cv->evaluate();
+
+		get_global_parallel()->set_num_threads(4);
```

Comment:
this is a by-chance checking of the consistency... depends on many aspects. for example on the machine that the code is being run on etc.

Reply:
true, but this already revealed bugs. open for ways to improve

Reply:
that's one thing that you have a test locally to have this tested... its another thing that have this formalised in a test that is part of the codebase and the CI that is going to be run on many different arch/distro etc...

Reply:
for example:
 * do not use global_parallel - that is going to be dropped eventually - the sooner the better
 * why 4, why not 8, or 3 or 11 or 42? you could actually use a function to get the runtime env supported number of max threads and use that number, and if it's 1 then say that foobar.
....

Reply:
the number is of course arbitrary, good idea to use the available number of threads, i will do that